### DISCUSSION TITLE
Custom Model for Stock Notifications: How to Notify?

### DISCUSSION THREAD

**Discussion by coderreco** [swellstores/community#571] [Help] [answered]
Working on a custom setup to allow customers to be notified when a product returns to in stock. Still planning but would like to implement this week.

As far as I have figured out so far:
1. Create a custom model called Stock Notifications
Fields:
```
stock-notification:
  - product_id (Text): The ID of the product
  - variant_id (Text): The ID of the specific variant (if applicable)
  - email (Text): The customer's email address
  - created_at (Date): The date the subscription was created
```

2. Customer subscribes to a stock notification on the frontend, creates a new record in the stock-notification model
3. Create an API endpoint that will get periodically hit by a cron job. Endpoint checks stock levels and for stock-notification records that's products are now in stock.
4. Send email notifications to the subscribers.
5. Delete the records.

Sending the notifications is where I am looking for guidance. I have read the docs but have not yet created a custom notification in Swell and would love to be pushed in the right direction. If using conditions I imagine I would associate this with the product model, and when a record updates check the stock level for greater than 0? Or would it be easier just to use a webhook to the endpoint when a `product.stock_adjusted` event occurs, then use the id of that product to get `stock-notification` records, and trigger the notification manually using the emails from those `stock-notification` records?

Thanks in advance!

**Comment by logeshswell** [comment, marked as answer]
@coderreco This looks like an excellent Swell app idea. Do you want to try it?

> Sending the notifications is where I am looking for guidance

https://developers.swell.is/guides/notifications

I would recommend following the below steps in addition to your thoughts:
1. Create a custom event (products.restocked) in stocks collection with the right conditions.
2. Trigger an async task (or webhook) when the above event is triggered from Swell.
3. Create a custom notification associated with stock-notification model and trigger an email using `$notify` operator inside your webhook implementation.

On a side note, add an additional field (`unsubscribed`) in stock-notification model to stop sending an email to unsubscribers.

These can be achieved easily via Swell apps. If you are interested, please check this out:
https://github.com/orgs/swellstores/discussions/458

**Reply by coderreco** [reply]
Would love to. Submitted a request to join the alpha a couple weeks ago but I haven't heard back.

**Reply by sags95** [reply]
Hey @coderreco, sending you a quick message now, looks like your store is on the Community plan. As Apps require a test environment we'll have to setup a new store for testing.

**Reply by coderreco** [reply]
Hey @sags95. The store that this will be implemented on is on another plan(Standard Plan I think?) so I can give you that store id instead

**Reply by sags95** [reply]
Yes that's correct, I can enable Apps access on any store that has the multi-environments feature. I sent you an email a few hours ago so feel free to contact me through there!

**Reply by coderreco** [reply]
Just saw! Will be in touch with store info soon!

Thanks @sags95

**Reply by coderreco** [reply]
Finally getting around to working on this but getting hung up on the notification portion. Anyone I can share code with for some guidance?

**Reply by ericingram** [reply]
In order to manually trigger a notification on a model, it looks like this
```
swell.put('/stock-notifications'/{id}, {
  id,
  $notify: '<name_of_notification_template>'
})
```

It can also trigger automatically based on the conditions of the record, for example if you had a field on that model called `back_in_stock` then you could configure conditions of the notification like this
```
{
  ...
  conditions: { back_in_stock: true }
}
```

**Reply by coderreco** [reply]
Hey @ericingram.

I am trying to trigger the notification on my `stock-notifications` model using the `product.updated` event.

Code:
```
// /models/stock-notifications.json
{
  "collection": "stock-notifications",
  "label": "Stock Notifications",
  "fields": {
    "product_id": {
      "type": "objectid",
      "required": true
    },
    "product": {
      "type": "link",
      "model": "products",
      "key": "product_id"
    },
    "variant_id": {
      "type": "objectid"
    },
    "variant": {
      "type": "link",
      "model": "products.variants",
      "key": "variant_id"
    },
    "email": {
      "type": "string",
      "required": true
    },
    "unsubscribed": {
      "type": "boolean",
      "default": false
    }
  }
}
```
--
```
// /content/stock-notifications.json
{
  "collection": "stock-notifications",
  "label": "Stock Notifications",
  "title": "Stock Notification",
  "description": "Manage customer notifications for out-of-stock products",
  "fields": [
    {
      "id": "product",
      "type": "product_lookup",
      "label": "Product",
      "required": true
    },
    {
      "id": "variant",
      "type": "variant_lookup",
      "label": "Variant"
    },
    {
      "id": "email",
      "type": "email",
      "label": "Customer Email",
      "required": true
    },
    {
      "id": "unsubscribed",
      "type": "toggle",
      "label": "Unsubscribed",
      "default": false
    }
  ],
  "views": [
    {
      "id": "list",
      "title": "Stock Notifications",
      "nav": {
        "parent": "products",
        "label": "Stock Notifications"
      },
      "fields": [{ "id": "email" }, { "id": "unsubscribed" }],
      "tabs": [
        {
          "id": "all",
          "label": "All"
        },
        {
          "id": "active",
          "label": "Active",
          "query": { "unsubscribed": false }
        },
        {
          "id": "inactive",
          "label": "Inactive",
          "query": { "unsubscribed": true }
        }
      ]
    },
    {
      "id": "new",
      "title": "New Stock Notification",
      "fields": [
        { "id": "product" },
        { "id": "variant" },
        { "id": "email" },
        { "id": "unsubscribed" }
      ]
    },
    {
      "id": "edit",
      "title": "Edit Stock Notification",
      "fields": [
        { "id": "product" },
        { "id": "variant" },
        { "id": "email" },
        { "id": "unsubscribed" }
      ]
    }
  ]
}
```
--
```
// /notifications/stock-notifications.json
{
  "collection": "stock-notifications",
  "label": "Product Back in Stock",
  "description": "Notify customers when a product they're interested in is back in stock",
  "method": "email",
  "subject": "{{ product.name }} is back in stock!",
  "contact": "email",
  "conditions": {
    "$and": [
      {
        "product_id": {
          "$eq": "{{ product.id }}"
        }
      },
      {
        "unsubscribed": {
          "$eq": false
        }
      }
    ]
  },
  "query": {
    "expand": ["product", "variant"]
  },
  "event": "product.updated",
  "new": false,
  "repeat": false,
  "admin": false
}
```
--
```
// /notifications/stock-notification.tpl
<!DOCTYPE html>
<html lang="en">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Product Back in Stock</title>
</head>
<body>
  <h1>Good news! A product you were interested in is back in stock.</h1>
  <p>The following product is now available:</p>
  <h2>{{ product.name }}</h2>
  <p>Price: {{ product.price }} CAD</p>
  <a href="{{ store.url }}/products/{{ product.slug }}">View Product</a>
</body>
</html>
```
This is where I am at. What I want to know is if this is the best way of doing it for this use case and if so can you see what I am missing? Currently not successfully getting notifications when testing. Instead of this notification setup should I be using a webhook or a function or combination of both? I also need to update the records that get notified to `unsubscribed: true` so I will likely use a function for that. Perhaps it is more efficient to combine both of these processes into one function?

Feels like I am close, and the dx so far is pretty good. Looking forward to working on some other apps later on.

Thanks in advance for your help!

**Reply by ericingram** [reply]
From here, the main question is how do you plan on triggering the back-in-stock condition? Is it a webhook based on product stock or some external system?

**Reply by coderreco** [reply]
@ericingram What I would like to do is trigger everything off of the product model using the `product.updated` event.

Important parts:

1. Product inventory is updated
2. Notifications are to be sent to records in the `stock-notifications` collection with conditions for records where `unsubscribed: false` and `product_id: {"$eq": "{{ product.id }}"}`
3. Records that are notified that the product stock level has changed are updated to `unsubscribed: true`

So what I want to know is should this notification be working(it isn't right now but could be due to something else), and if I should instead be using a separate approach?
```
// /notifications/stock-notifications.json
{
  "collection": "stock-notifications",
  "event": "product.updated",
  "conditions": {
    "$and": [
      {
        "product_id": {
          "$eq": "{{ product.id }}"
        }
      },
      {
        "unsubscribed": {
          "$eq": false
        }
      }
    ]
  },
  "query": {
    "expand": ["product", "variant"]
  },
  "label": "Product Back in Stock",
  "description": "Notify customers when a product they're interested in is back in stock",
  "method": "email",
  "subject": "{{ product.name }} is back in stock!",
   "contact": "email",
}
```
I could use a function or a function and a custom event that only emits when the `stock_level` has changed, I just want to make sure I am notifying all of the `stock-notification` records that qualify because there could be 1, or there could be 101. 

If I want to handle all of this through swell, what do you recommend I do?

**Reply by coderreco** [reply]
Checking back in on this

**Reply by ericingram** [reply]
I would recommend making a custom event and using a function to trigger the stock notifications. You just need to iterate over them with something like this:
```
const stockNotifications = await swell.get('/stock-notifications', {
  product_id: <relevant_product_id>,
  unsubscribed: { $ne: true },
  limit: 100 // consider using pagination
})

for (const stockNote of notifications.results) {
  swell.put('/stock-notifications/{id}', {
    id: stockNote.id,
    $notify: 'product-back-in-stock' // or whatever the name of the notification is
  })
}
```

**Reply by coderreco** [reply]
I'll try that, thanks!

**Reply by coderreco** [reply]
Alright, I am mostly there but hitting a couple of issues getting the notification itself to send the email. I do have the notifications correctly firing on the model using `$notify`, but when I look at `/notifications` in the console I am seeing the following:
`"error": "Template content not defined"`

I do have a liquid template created using the same name as the notification configuration file, I used the honest-reviews-app as a reference to see if I could find what I am missing but I cannot. What do I need to do to connect the html content to the notification?

```
{
  "collection": "restock-notices",
  "label": "Product Back in Stock",
  "description": "Notify customers when a product they're interested in is back in stock",
  "method": "email",
  "subject": "{{ product.name }} is back in stock!",
  "contact": "email",
  "query": {
    "expand": ["product", "variant"]
  },
  "data": {
    "product": "{{ product }}",
    "variant": "{{ variant }}",
    "email": "{{ email }}"
  }
}
```
```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Product Back in Stock</title>
</head>
<body>
  <h1>Good news! A product you were interested in is back in stock.</h1>
  <p>The following product is now available:</p>
  <h2>{{ product.name }}</h2>
  <p>Price: {{ product.price }} CAD</p>
  <a href="{{ store.url }}/products/{{ product.slug }}">View Product</a>
</body>
</html>
```

**Reply by ericingram** [reply]
Can you share a zip of the app with me? eric@swell.is

**Reply by ericingram** [reply]
@coderreco Thanks for sending it. You're right that I don't see the notification template getting created with the app.

We'll look deeper and find out why

**Reply by ericingram** [reply]
@coderreco we found a bug in the way notification templates are pushed to the API. A fix should be released soon, that will require updating your `@swell/cli` version also.

**Reply by ericingram** [reply]
@coderreco a fix was made, try it again using CLI version `2.0.2-alpha.0`

**Reply by coderreco** [reply]
Still not working for me. Do I need to version the app or anything? Doesn't seem like any files are updating.

**Reply by ericingram** [reply]
Try making a small change to the notification files. They won't re-push if the backend thinks they didn't change
